### Setting

In [15]:
import glob
import shutil
import os
import random

input_dir = "/opt/ml/input"
data_dir = os.path.join(input_dir, "data")
sample_dir = os.path.join(input_dir, "sample_data")

categories = ["Battery", "Clothing", "Glass", "Metal", "Paper", "Paperpack", "Plastic", "Plasticbag", "Styrofoam"]

In [17]:
# mask sample_data directory
os.makedirs(sample_dir, exist_ok=True)
os.makedirs(os.path.join(sample_dir, "test", "NoLabel"), exist_ok=True)

for category in categories:
    os.makedirs(os.path.join(sample_dir, "train", category), exist_ok=True)
    os.makedirs(os.path.join(sample_dir, "val", category), exist_ok=True)

### 데이터 개수 확인

In [18]:
print(f"{'Category':10} | train |   val | total")
print("--------------------------------------")
for category in categories:
    src_dir = os.path.join(data_dir, "train", category)
    train_imgs = glob.glob(os.path.join(src_dir, "*"))
    
    src_dir = os.path.join(data_dir, "val", category)
    val_imgs = glob.glob(os.path.join(src_dir, "*"))

    print(f"{category:10} | {len(train_imgs):5} | {len(val_imgs):5} | {len(train_imgs)+len(val_imgs)}")

Category   | train |   val | total
--------------------------------------
Battery    |   261 |    87 | 348
Clothing   |   676 |   225 | 901
Glass      |  1169 |   389 | 1558
Metal      |  1669 |   556 | 2225
Paper      |  6895 |  2298 | 9193
Paperpack  |  1458 |   486 | 1944
Plastic    |  3793 |  1264 | 5057
Plasticbag |  6970 |  2323 | 9293
Styrofoam  |  1560 |   520 | 2080


### 일부 데이터(sample data) 복사
- `/opt/ml/input/data` -> `/opt/ml/input/sample_data` 폴더
- train data
    - Battery  : 모두(261 개) 복사
    - Clothing : 모두(676 개) 복사
    - 그 외 : 1000개 복사
- val data
    - 복사한 train data의 30% 복사
- test data는 모두 복사

In [19]:
# copy test data (all)

src_dir = os.path.join(data_dir, "test", "NoLabel")
dst_dir = os.path.join(sample_dir, "test", "NoLabel")
for jpgfile in glob.iglob(os.path.join(src_dir, "*")):
    shutil.copy(jpgfile, dst_dir)

print("Done.")

Done.


In [20]:
# copy train data

num = 100  # Battery, Clothing 외의 데이터 복사할 개수
for category in categories:
    # copy train data
    src_dir = os.path.join(data_dir, "train", category)
    dst_dir = os.path.join(sample_dir, "train", category)
        
    data = glob.glob(os.path.join(src_dir, "*"))
    # random.shuffle(data)
    for jpgfile in data[:num]:
        shutil.copy(jpgfile, dst_dir)

print("Done.")

Done.


In [21]:
# copy validation data

times = 0.3  # train data의 30% 복사

for category in categories:
    # copy validation data
    sample_train_dir = os.path.join(sample_dir, "train", category)
    src_dir = os.path.join(data_dir, "val", category)
    dst_dir = os.path.join(sample_dir, "val", category)
    
    sample_train_data = glob.glob(os.path.join(sample_train_dir, "*"))
    data = glob.glob(os.path.join(src_dir, "*"))
    # random.shuffle(data)
    num = int(len(sample_train_data)*times)
    for jpgfile in data[:num]:
        shutil.copy(jpgfile, dst_dir)
        
print("Done.")

Done.


### 복사된 데이터(sample data) 개수 확인

In [22]:
print(f"{'category':10} | train | validation")
print(f"{'-'*30}")

for category in categories:
    category_dir = os.path.join(sample_dir, "train", category)
    sample_data = glob.glob(os.path.join(category_dir, "*"))
    num = len(sample_data)
    print(f"{category:10} | {num:4}  |", end="")
    
    category_dir = os.path.join(sample_dir, "val", category)
    sample_data = glob.glob(os.path.join(category_dir, "*"))
    num = len(sample_data)
    print(f"{num:5}")

category   | train | validation
------------------------------
Battery    |  100  |   30
Clothing   |  100  |   30
Glass      |  100  |   30
Metal      |  100  |   30
Paper      |  100  |   30
Paperpack  |  100  |   30
Plastic    |  100  |   30
Plasticbag |  100  |   30
Styrofoam  |  100  |   30


## 복사된 데이터(sample data) 일부 파일 명 확인

In [23]:
# Check train data
print(f"{'train':^33} | {'validation':^30}")
print(f"{'-'*66}")

num = 10  # 각 데이터 10개씩만 확인
for category in categories:
    print(category)
    category_dir = os.path.join(sample_dir, "train", category)
    sample_train_data = glob.glob(os.path.join(category_dir, "*"))
    
    category_dir = os.path.join(sample_dir, "val", category)
    sample_val_data = glob.glob(os.path.join(category_dir, "*"))
    for i, (train_img, val_img) in enumerate(zip(sample_train_data[:num], sample_val_data[:num])):
        train_fn = train_img.split('/')[-1]
        val_fn = val_img.split('/')[-1]
        print(f"{i:3}: {train_fn} | {val_fn}")
    print(f"{'-'*66}")

              train               |           validation          
------------------------------------------------------------------
Battery
  0: jvcbfelzcxnwlcvdqicgiawf.jpg | 0kgixtccxvclspcodgvrssfv.jpg
  1: 00lgniecqhkrwtmygiekbkoq.jpg | 0000zvojloffsgodgfzdhsbx.jpg
  2: 000hnrcodsgzhvpyouuiwafr.jpg | 000krbyibuioecrdqvnoopkv.jpg
  3: 0ktbphrszirwyjktuizczmvq.jpg | kltcfdzezyhokqdkiswtovoj.jpg
  4: pcqakahlvgqbbxsddzmecruf.jpg | gdvgqbapduhaljdmpyvvtmne.jpg
  5: 00oyhkqmybaxudmkfmlpjggo.jpg | 000vahrzpvvgvuptnhynfnxx.jpg
  6: 00rzpdqvjlakhwnocbyizyka.jpg | rjgpwlweyanhkyxwymzelmly.jpg
  7: 0000rdmgxsvlxjmpiuklwpmq.jpg | 0000hmlckscazzfjduiucive.jpg
  8: 00mttlodgsekxrlbgktlymeg.jpg | 00yvfezewqcxcwtpewvyiiwu.jpg
  9: 0000mnlgsibktntcmcelpuze.jpg | 0000ygpivjeqbkvbrtibslav.jpg
------------------------------------------------------------------
Clothing
  0: 00ijauedmmlzsncfyrrsduuo.jpg | 00mbrcoofiwoxeldfmqkrrsj.jpg
  1: 00fjyomqiahizmeogbponnln.jpg | 0000usznmlqtrziqcbnndxcm.jpg
  